### PR TITLE
connect only to primary host for load balancer scenario #999

### PR DESCRIPTION
**Pull request by ehennum**
To create a DataMovementManager for use with a load balancer, the factory call should be:

    newDataMovementManager(DatabaseClient.ConnectionPolicy.PRIMARY_HOST)

No FilteredForestConfiguration can be used when connecting only to the primary host.  Thus, no whitelist should be possible or needed.  The builtin HostAvailabilityListener and NoResponseListener should work in failover situations.

Please consider whether the design changes are usable and test in an ALB environment.

If the changes are good, please merge into 4.1.1-develop.

### CODE REVIEW

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/HostAvailabilityListener.java [hunk 1]
```diff
@@ -129,15 +132,22 @@ public HostAvailabilityListener withSuspendTimeForHostUnavailable(Duration
    * @return this instance (for method chaining)
    */
   public HostAvailabilityListener withMinHosts(int numHosts) {
-    if (numHosts <= 0) throw new IllegalArgumentException("numHosts must be > 0");
-    int numConfigHosts = moveMgr.readForestConfig().getPreferredHosts().length;
-    if (numHosts > numConfigHosts) throw new IllegalArgumentException
-      ("numHosts must be less than or equal to the number of hosts in the cluster");
+    if (moveMgr.getConnectionPolicy() == DatabaseClient.ConnectionPolicy.PRIMARY_HOST) {
+      if (numHosts != 0) {
+        throw new IllegalArgumentException("numHosts must be 1 when using only the primary host for the connection");
```
Comment (srinathgit) [inline marklogic-client-api/src/main/java/com/marklogic/client/datamovement/HostAvailabilityListener.java]: Shouldn't the condition be  if (numHosts != 1) ?

**Comment by srinathgit** [comment]
@ehennum ,

So, in case where a MarkLogic host is down, the user will still have to remove 'HostAvailabilityListener' and 'NoResponseListener' and add their custom listener based on what exception they get with their LoadBalancer ?

File: marklogic-client-api/src/main/java/com/marklogic/client/datamovement/HostAvailabilityListener.java (hunk 1 above)
Comment (ehennum) [reply marklogic-client-api/src/main/java/com/marklogic/client/datamovement/HostAvailabilityListener.java]: Yes, good catch.

**Comment by ehennum** [comment]
@srinathgit, no, it should no longer be necessary to remove HostAvailabilityListener or NoResponseListener.

The opposite:  to detect a failure of the load balancer, adopters will need to add a custom listener for the specific errors generated by their preferred load balancer.

Possibly, as a failsafe in the primary host case, the batcher should keep a count of the number of retry efforts on the job and fail if it exceeds a limit.

**Comment by srinathgit** [comment]
@ehennum , In a 3 node AWS cluster with an ALB,  I ran a write batcher job , during the job if MarkLogic instance is shutdown (to force a forest failover ) and a request is sent to that host, the "FailedRequestException" is thrown which is not part of the "hostUnavailableExceptions" list. There is a method "withHostUnavailableExceptions" where we can specify the list of "host unavailable exceptions" however "FailedRequestException" can't be added to the list as it is too generic and could be thrown for other valid reasons.  So, in this scenario neither the job gets shut down nor the failed batches retried.
```
19:38:35.287 [pool-2-thread-20] DEBUG c.m.client.impl.OkHttpServices - Sending multipart for /v1/documents
com.marklogic.client.FailedRequestException: Local message: failed to apply resource at documents: Bad Gateway. Server Message: Server (not a REST instance?) did not respond with an expected REST Error message.
	at com.marklogic.client.impl.OkHttpServices.checkStatus(OkHttpServices.java:4327)
	at com.marklogic.client.impl.OkHttpServices.postResource(OkHttpServices.java:3400)
	at com.marklogic.client.impl.OkHttpServices.postBulkDocuments(OkHttpServices.java:3480)
	at com.marklogic.client.impl.DocumentManagerImpl.write(DocumentManagerImpl.java:614)
	at com.marklogic.client.impl.GenericDocumentImpl.write(GenericDocumentImpl.java:1)
	at com.marklogic.client.impl.DocumentManagerImpl.write(DocumentManagerImpl.java:606)
	at com.marklogic.client.impl.GenericDocumentImpl.write(GenericDocumentImpl.java:1)
	at com.marklogic.client.datamovement.impl.WriteBatcherImpl$BatchWriter.run(WriteBatcherImpl.java:1060)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

In an on-prem 3 node MarkLogic cluster, during the same job, if a MarkLogic instance is shutdown ,  java.net.SocketException is thrown which is part of the "hostUnavailableExceptions" list. 
```
20:20:34.074 [pool-2-thread-8] ERROR c.m.c.d.HostAvailabilityListener - ERROR: host unavailable "rh7v-intel64-90-test-18.marklogic.com", black-listing it for PT10M
com.marklogic.client.MarkLogicIOException: java.net.SocketException: Connection reset
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:708)
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:700)
	at com.marklogic.client.impl.OkHttpServices.doPost(OkHttpServices.java:4071)
	at com.marklogic.client.impl.OkHttpServices.postResource(OkHttpServices.java:3372)
	at com.marklogic.client.impl.OkHttpServices.postBulkDocuments(OkHttpServices.java:3480)
	at com.marklogic.client.impl.DocumentManagerImpl.write(DocumentManagerImpl.java:614)
	at com.marklogic.client.impl.GenericDocumentImpl.write(GenericDocumentImpl.java:1)
	at com.marklogic.client.impl.DocumentManagerImpl.write(DocumentManagerImpl.java:606)
	at com.marklogic.client.impl.GenericDocumentImpl.write(GenericDocumentImpl.java:1)
	at com.marklogic.client.datamovement.impl.WriteBatcherImpl$BatchWriter.run(WriteBatcherImpl.java:1060)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at okio.Okio$2.read(Okio.java:140)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:355)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:227)
	at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215)
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.burgstaller.okhttp.AuthenticationCacheInterceptor.intercept(AuthenticationCacheInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at com.marklogic.client.impl.OkHttpServices.sendRequestOnce(OkHttpServices.java:706)
	... 14 common frames omitted
```

**Comment by ehennum** [comment]
@srinathgit , thanks for doing the comparison.  It sounds like the Java Client API receives a different HTTP response in the ALB scenario.   If so, can you report the difference in the HTTP response that causes the same server event to throw a FailedRequestException instead of a SocketException?

The only change in the HostAvailabilityListener behavior in the primary host case is that it skips the host refresh that culminates in the invalidation of retry on this line:

https://github.com/marklogic/java-client-api/blob/431c1a273b7f6b5c87cadd576fc600ca436c42e2/marklogic-client-api/src/main/java/com/marklogic/client/datamovement/HostAvailabilityListener.java#L299

**Comment by srinathgit** [comment]
@ehennum , I printed the HTTP response for the ALB scenario and after the instance is stopped, the following is the response I get :

Response{protocol=http/1.1, code=502, message=Bad Gateway, url=http://srinath-q-elasticl-a1rzued8pwa4-882617851.us-east-2.elb.amazonaws.com:8008/v1/documents}

Response body:
<html>
<head><title>502 Bad Gateway</title></head>
<body bgcolor="white">
<center><h1>502 Bad Gateway</h1></center>
</body>
</html>

**Comment by srinathgit** [comment]
@ehennum ,

I spoke with @mattsunsjf  , 

1. The ELB can send only "HTTP code=502, message=Bad Gateway" response for requests to the MarkLogic instance that was shutdown. It cannot be pass through and send the response it got to the client
2. Regarding requests getting sent to the downed instance, Matt will open a task in JIRA and will look further to see what the right behavior is. In my case, I had downed the instance that had the "Security" db so I was getting "Bad Gateway" for every request. Nevertheless,  even if the instance that doesn't have "Security" db is taken down, there were cases where requests that were sent to it.

**Pull request merged**